### ISSUE TITLE
Develop a task-specific grounded semantic proposer for REQUIRE

### ISSUE THREAD

**Issue by grantaj** [grantaj/thorn#222]
## Summary

#219 established a clean architectural result:

> Off-the-shelf zero-shot relation extraction preserves Thorn's supplied endpoint/provenance boundary, but does not recover natural mathematical prerequisite semantics with useful coverage.

The frozen public GLiREL configuration achieved a small fail-closed region on the synthetic benchmark, and the subsequent untouched private qualification established the natural zero-shot baseline at **3/31 REQUIRE occurrences recovered (9.7% recall)**. The private negative denominator was too small to establish natural precision. GLiNER-RelEx had substantially worse false-authority behavior on the public benchmark.

The next question is therefore no longer whether another handwritten compiler or zero-shot label wording can solve `REQUIRE`. It is:

> **Can task-specific supervision adapt an existing local language encoder/relation extractor to identify direct mathematical prerequisite use, over Thorn-supplied entities and exact source context, while preserving fail-closed abstention and exact provenance?**

This issue defines that public research programme and its model-neutral contract. It is **not** a production-integration issue and does not authorize training on the existing private #8/#12 holdout.

## Architectural boundary

Keep the #212 distinction explicit:

```text
Resolve(reference) != REQUIRE(owner, resolved_target)
```

Thorn independently supplies or establishes:

- the current dependent result/proof owner `u`;
- the exact result-reference occurrence `r`;
- occurrence identity and exact source provenance;
- independent reference resolution `Resolve(r) = v` where uniquely available;
- bounded source context / reversible linguistic projection;
- generic linguistic/source facts where useful.

The learned component may only interpret whether the supplied occurrence expresses direct prerequisite use.

Conceptually:

```text
exact source + typed occurrences + linguistic facts
                    |
                    v
       MathematicalSemanticProposer
                    |
        grounded REQUIRE proposal
        or explicit abstention
                    |
                    v
        Thorn authority/provenance gate
                    |
           canonical REQUIRE(u, v)
           or unresolved
```

The model **does not decide mathematical authority**.

## Required research interface

Preserve the model-neutral contract introduced under #219, generalized for trainable candidates:

```text
RequireCase(
    owner,
    context,
    reference_occurrence,
    resolved_target,
    provenance,
    optional_structural_or_linguistic_features,
)

-> RequireProposal(
    score,
    exact_reference_occurrence,
    optional_evidence_span,
)
```

Decision semantics remain asymmetric:

- a sufficiently high score may propose `REQUIRE`;
- a lower score means **abstain / unresolved**;
- low score must never assert canonical `NON_REQUIRE`;
- the model must not invent an owner, endpoint, source span, or resolved target.

Where a candidate architecture naturally produces `REQUIRE` / `NON_REQUIRE` / `UNRESOLVED` classes internally, Thorn still treats non-`REQUIRE` output as absence of authority rather than proof of non-dependence.

## Existing frozen baseline

The following evidence is now historical/frozen and must remain comparable:

### Public #219 synthetic benchmark

- 34 cases / 39 exact reference occurrences;
- positive, resolved-reference negative, ambiguous, multi-reference, cross-sentence and operator-absent cases;
- exact endpoint preservation measured separately from relation correctness.

Selected zero-shot GLiREL operating point:

- `jackboyla/glirel-large-v0`;
- frozen model revision `40a523e12a8432d6da364cf2a195a28755ff04d3`;
- label `is required to establish the current result`;
- threshold `0.20`;
- public TP 3 / FP 0 / FN 14;
- precision 1.0, recall 0.1765, ambiguous assertions 0/7.

### Private natural zero-shot baseline

Aggregate non-identifying result only:

- 31 annotated natural `REQUIRE` occurrences;
- TP 3 / FN 28 at the exact frozen public configuration;
- recall **0.0968**;
- exact endpoint transport 33/33;
- one explicit negative and one fail-closed ambiguous control were not asserted;
- insufficient natural negative denominator to establish precision.

The existing private cases are **not a development set**.

## Corpus separation / contamination rule

The private #8/#12 natural cases are permanently excluded from:

- training;
- fine-tuning;
- threshold selection;
- early stopping;
- relation-label or prompt selection;
- context-window selection;
- feature selection;
- hyperparameter selection;
- error-driven rule/model design.

They may remain as historical frozen baseline evidence only.

A new private development corpus must be constructed under a separate issue from different papers. Development/training and final qualification must be split by **paper**, not by sentence/reference occurrence, to avoid author/style/content leakage.

Final qualification of a tuned candidate must use a **fresh natural-paper holdout that was never used for development**.

## Candidate technology

Do **not** train a language model from scratch in the first tranche.

Evaluate adaptation of mature pretrained local models behind the same contract. At minimum consider:

1. **GLiREL fine-tuning/adaptation**
   - strong architectural fit because supplied entities already work;
   - directly comparable with the zero-shot baseline;
   - useful test of whether task-specific relation supervision is sufficient.

2. **A conventional pretrained encoder with a small task-specific relation/classification head**
   - e.g. a current BERT-family scientific/general encoder suitable for local fine-tuning;
   - owner/reference markers and exact candidate positions supplied explicitly;
   - no generative output required.

A third candidate may be justified if it tests a materially different hypothesis, but avoid a broad model zoo.

Model choice is empirical. Do not make GLiREL a production dependency merely because it is the first baseline.

## Training target

The target is intentionally small:

> For one exact supplied result-reference occurrence in bounded mathematical context, estimate whether that occurrence expresses a **direct prerequisite of the current result**.

Training data should preserve hard distinctions exposed by #212/#219:

- direct theorem/result use;
- mention/background;
- comparison/contrast;
- attribution/history;
- explicit non-use;
- hypothetical/conditional use;
- rhetorical `see` / `recall` / `compare`;
- multi-reference sentences with mixed roles;
- joint support;
- alternative support / ambiguity;
- cross-sentence support;
- operator-present and operator-absent expressions;
- proof and theorem-statement contexts.

Do not create a hidden synonym dictionary as preprocessing. Generic token/POS/dependency/source facts may be supplied as features if they are mechanically derived and independently testable.

### Development metrics

Report at least:

- precision / recall / F1 on `REQUIRE`;
- precision-versus-coverage curve under abstention;
- false-`REQUIRE` rate over hard resolved-reference negatives;
- ambiguous-case assertion rate;
- exact endpoint/provenance preservation;
- operator-present versus operator-absent recall;
- multi-reference discrimination;
- cross-sentence performance;
- paper-level generalization;
- calibration / score reliability where meaningful;
- runtime, peak memory, model size and CPU feasibility.

Do not report accuracy as the principal metric if class balance makes it misleading.

### Safety-first advancement gate

Before consuming a fresh final natural holdout, a candidate should demonstrate on paper-separated development evaluation:

- exact endpoint/provenance transport **100%**;
- no known path for model output to bypass the Thorn authority gate;
- `REQUIRE` precision at least **0.98** at the selected abstention threshold, with a sufficiently large negative set that this is not supported by one or two examples;
- `REQUIRE` recall at least **0.50** at that same operating point, materially above the 9.7% frozen natural zero-shot baseline;
- ambiguous cases predominantly abstained rather than asserted;
- no catastrophic subtype such as comparison/non-use being systematically promoted;
- selected model/checkpoint/threshold/context/features frozen before final holdout.

These are **research progression gates**, not a claim that 0.98 precision is automatically sufficient for production mathematical authority. Production qualification, if this programme succeeds, is a later issue.

If no candidate can approach a useful precision/coverage regime, stop and record the result rather than widening authority or adding paper-specific rules.

### Tranche A — training/evaluation substrate

- consume a new private paper-separated development corpus through a stable research interface;
- keep private source/material out of the public repo;
- support deterministic training manifests and frozen split identities;
- keep the existing public #219 benchmark runnable as a regression/comparator;
- make model training local/keyless where practical;
- record exact pretrained model/revision, package versions, seed(s), hyperparameters and dataset identity.

### Tranche B — small supervised comparison

Train at least the two candidate families above under comparable splits/metrics.

The first objective is not maximum leaderboard performance. It is to determine whether task-specific supervision produces a **qualitatively better safety/coverage frontier** than zero-shot extraction.

### Tranche C — freeze one candidate

Only if a candidate clears the development gate:

- freeze exact weights/checkpoint hash;
- freeze preprocessing/tokenization/context/features;
- freeze threshold/calibration;
- freeze runtime/package identity;
- prohibit further development against final holdout.

### Tranche D — fresh private qualification

Use a fresh paper-level holdout under a separate private issue. Do not reuse #8/#12 as the final tuned-model test set.

## Production guard

This issue is research-only.

Do not:

- change production `DependencyGraph` semantics;
- promote learned scores directly to canonical `REQUIRE`;
- change `Resolve` / `Direct` semantics;
- alter #212 graph calculus;
- reintroduce the handwritten English compiler as authority;
- add paper-specific rules based on corpus errors;
- mix `DECLARE` into this first trainable tranche;
- train on or tune against the private #8/#12 holdout;
- make a large model a mandatory Thorn runtime dependency before qualification.

The learned component interprets evidence. **Thorn decides authority.**

## Stop conditions

Stop and record a negative result if:

- task-specific tuning still cannot produce a useful high-precision coverage regime;
- success depends on author/paper-specific lexical templates;
- exact endpoint/source grounding cannot be guaranteed;
- model confidence cannot support conservative abstention;
- a simpler model performs equivalently and the larger system adds no value;
- measured improvements disappear under paper-level held-out evaluation.

A negative result here is architecturally important: it would indicate that the remaining semantic interpretation problem may require a substantially stronger component or a different product boundary.

## Acceptance criteria

- [ ] A model-neutral trainable `REQUIRE` proposal interface is documented and executable.
- [ ] Existing #219 zero-shot public results remain reproducible and unchanged.
- [ ] Private #8/#12 cases are explicitly excluded from all development/tuning paths.
- [ ] Development and evaluation split at paper level, not occurrence level.
- [ ] At least GLiREL adaptation and one conventional encoder classifier are compared.
- [ ] No language model is trained from scratch in the first tranche.
- [ ] Precision/coverage, false authority, ambiguity, endpoint grounding and paper-level generalization are first-class metrics.
- [ ] Exact source/provenance identity remains supplied and mechanically checked.
- [ ] Model output remains non-authoritative until an independent Thorn gate.
- [ ] Candidate weights/configuration/threshold are frozen before any fresh final natural holdout.
- [ ] A final disposition states whether task-specific supervision makes grounded `REQUIRE` extraction plausibly viable.
- [ ] No production integration occurs under this issue.

## Success criterion

A successful research outcome should support the statement:

> **Using Thorn-supplied exact result/reference identities and bounded mathematical context, a task-specific local model can recover a materially useful subset of direct prerequisite relations at very high precision, abstain safely on uncertainty, and preserve exact provenance—without Thorn hand-compiling mathematical English.**
